source/h1_connection.c

```diff
@@ -923,6 +940,13 @@ static void s_outgoing_stream_task(struct aws_channel_task *task, void *arg, enu
         aws_mem_release(msg->allocator, msg);
     }
 
+    /* The rest of the pending chunks will be taken care of when the http stream is destroyed.
+     * Only need to clean up the current one being processed here.*/
+    struct aws_http1_stream_chunk *cur_chunk = connection->thread_data.encoder.message->body_chunk;
+    if (cur_chunk) {
+        aws_h1_stream_release_chunk(cur_chunk);
```

Comment:
Good point - I hadn't considered the case where the connection was paused, the socket timed out, and then the clean up of the stream was invoked. I have moved all memory clean up into two places: 
1. in the happy case, the current chunk is cleaned up when the chunk transitions into into the "end" or "terminated" state. 
2. In the unhappy case, when the stream is cleaned up. 

I have added asserts to aws_h1_populate_current_stream_chunk to ensure that the caller has nulled out the current chunk before invoking that call which should help catch a programming error later on where the writer forgets to clean up and null out the current chunk before trying to get the next one.

I considered just letting the clean up happen in aws_h1_populate_current_stream_chunk(), but I decided not to do that as it leaves the chunk memory hanging around until the stream clean up or the next chunk is called. On units with memory pressure, this could cause the memory to hang around until the stream is cleaned up on the final terminate call, which doesn't seem ideal. The trade off is that the could would only have two places where memory gets freed instead of three at the expense of keeping some memory around longer.

Reply:
👍